loopd/daemon.go

```diff
@@ -38,72 +42,89 @@ type listenerCfg struct {
-func daemon(config *Config, lisCfg *listenerCfg) error {
-	lnd, err := lisCfg.getLnd(config.Network, config.Lnd)
-	if err != nil {
-		return err
-	}
-	defer lnd.Close()
+// Daemon is the struct that holds one instance of the loop client daemon.
+type Daemon struct {
+	// To be used atomically. Declared first to optimize struct alignment.
+	started int32
+	stopped int32
 
-	// If no swap server is specified, use the default addresses for mainnet
-	// and testnet.
-	if config.SwapServer == "" {
-		// TODO(wilmer): Use onion service addresses when proxy is
-		// active.
-		switch config.Network {
-		case "mainnet":
-			config.SwapServer = mainnetServer
-		case "testnet":
-			config.SwapServer = testnetServer
-		default:
-			return errors.New("no swap server address specified")
-		}
-	}
+	// swapClientServer is the embedded RPC server that satisfies the client
+	// RPC interface. We embed this struct so the Daemon itself can be
+	// registered to an existing grpc.Server to run as a subserver in the
+	// same process.
+	swapClientServer
 
-	log.Infof("Swap server address: %v", config.SwapServer)
+	cfg             *Config
+	listenerCfg     *listenerCfg
+	mainErrChan     chan<- error
+	internalErrChan chan error
 
-	// Create an instance of the loop client library.
-	swapClient, cleanup, err := getClient(config, &lnd.LndServices)
-	if err != nil {
-		return err
+	lnd           *lndclient.GrpcLndServices
+	clientCleanup func()
+
+	wg sync.WaitGroup
+
+	mainCtx       context.Context
+	mainCtxCancel func()
+
+	grpcServer    *grpc.Server
+	grpcListener  net.Listener
+	restServer    *http.Server
+	restListener  net.Listener
+	restCtxCancel func()
+}
+
+// New creates a new instance of the loop client daemon.
+func New(config *Config, lisCfg *listenerCfg, errChan chan<- error) *Daemon {
+	return &Daemon{
+		cfg:         config,
+		listenerCfg: lisCfg,
+		mainErrChan: errChan,
+		// We have 3 goroutines that could potentially send an error.
+		// We react on the first error but in case more than one exits
+		// with an error we don't want them to block.
+		internalErrChan: make(chan error, 3),
 	}
-	defer cleanup()
+}
 
-	// Retrieve all currently existing swaps from the database.
-	swapsList, err := swapClient.FetchSwaps()
-	if err != nil {
-		return err
+// Start starts loopd in daemon mode. It will listen for grpc connections,
+// execute commands and pass back swap status information.
+func (d *Daemon) Start() error {
+	// There should be no reason to start the daemon twice. Therefore return
+	// an error if that's tried. This is mostly to guard against Start and
+	// StartAsSubserver both being called.
+	if atomic.AddInt32(&d.started, 1) != 1 {
+		return errOnlyStartOnce
 	}
 
-	swaps := make(map[lntypes.Hash]loop.SwapInfo)
-	for _, s := range swapsList {
-		swaps[s.SwapHash] = *s
+	var err error
+	d.lnd, err = d.listenerCfg.getLnd(d.cfg.Network, d.cfg.Lnd)
+	if err != nil {
+		return d.stopAndReturnErr(err)
```

Comment:
My suggestion was, instead of creating `stopAndReturnErr`, to create a wrapping method for `Start`. For example rename `Start` to `start` and add a new `Start` method that looks like:
```
err := start()
if err != nil {
   log(err)
   stop()
}
```

The advantage is that future devs can never forgot to call `stopAndReturnErr`

Reply:
Okay, that makes sense.